From e00ae3553d0fc3ced8d9807dd94130583abb38d3 Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Tue, 7 Oct 2025 01:17:57 +0200
Subject: [PATCH 03/20] win32u: Prevent osu! from stealing focus on certain
 window managers

Behavior can be toggled with WINE_ENABLE_OSU_FOCUS_FIX.
---
 dlls/win32u/window.c | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/dlls/win32u/window.c b/dlls/win32u/window.c
index b2f9b744622..b5e4dd81c1c 100644
--- a/dlls/win32u/window.c
+++ b/dlls/win32u/window.c
@@ -4006,6 +4006,28 @@ done:
     return after;
 }
 
+static BOOL osu_focus_hack()
+{
+    static char cache = -1;
+    if (cache == -1)
+    {
+        BOOL result = TRUE; 
+        const char *env = getenv("WINE_ENABLE_OSU_FOCUS_FIX");
+
+        /* Openbox and KDE have known issues with the hack */
+        if (user_driver->pHasWindowManager &&
+            (user_driver->pHasWindowManager("Openbox") || user_driver->pHasWindowManager("KWin")))
+            result = FALSE;
+
+        /* Also override behavior from env. var */
+        if (env)
+            result = atoi(env) != 0;
+
+        cache = result;
+    }
+    return !!cache;
+}
+
 /* NtUserSetWindowPos implementation */
 BOOL set_window_pos( WINDOWPOS *winpos, int parent_x, int parent_y )
 {
@@ -4017,6 +4039,13 @@ BOOL set_window_pos( WINDOWPOS *winpos, int parent_x, int parent_y )
 
     orig_flags = winpos->flags;
 
+    /* osu! hack: fix osu! taking focus over other apps on certain WMs */
+    if (osu_focus_hack() && winpos->hwndInsertAfter == HWND_NOTOPMOST && 
+        (get_window_long(winpos->hwnd, GWL_EXSTYLE) & WS_EX_TOPMOST))
+    {
+        winpos->flags |= SWP_NOACTIVATE | SWP_NOZORDER;
+    }
+
     /* First, check z-order arguments.  */
     if (!(winpos->flags & SWP_NOZORDER))
     {
-- 
2.55.0

